improvement(client): improved internal submitMessage, reSubmit, rollback, and submitSignal type-safety

jason-ha opened the pull request:
improvement(client): improved internal submitMessage, reSubmit, rollback, and submitSignal type-safety

In Legacy API:
- Add type parameter to `IEnvelope` for `contents` to support stricter typing where desired.
- Remove unused parameters from mock `submitMessage` and `submitSignal` methods.
- Clean up some mock types.

Internally:
- Replace `any` in `ISignalEnvelope` with `unknown`.
- Relocate `FluidDataStoreMessage` to runtime-definitions. Eventually this can serve as basis for message types for DataStores.
- In `datastore`, create `LocalFluidDataStoreRuntimeMessage` representing the two known messages for DataStores.
- `FluidDataStoreContext` requires `OutboundFluidDataStoreMessage` for stricter `reSubmit` and `rollback` methods.
- Create `IFluidRootParentContextPrivate` for the transition from DataStore to `ContainerRuntime` and use explicit types in `ContainerRuntime` and `ChannelCollection`.
- In `ChannelCollection` restore some old method names with strict typing that internal callers must conform to. The `IFluidDataStoreChannel` portions that were not required by current use were refactored to new `ComposableChannelCollection` class.
- Replace `wrapContext` with `formParentContext` that handles both `IFluidParentContextPrivate` and `IFluidRootParentContextPrivate` by requiring the specific `submitMessage` and `submitSignal` functions
- Within `ContainerRuntime`:
  - Update `submitMessage` to expect specific messages
  - Update `submitSignal` to expect new `AddressedUnsequencedSignalEnvelope` that is specific subset of `ISignalEnvelope` and is exactly `IEnvelope<ISignalEnvelope["contents"]>`
- Fixup tests to use more realistic data (as required by stricter typing).

Potential Future Improvements:
- Change {@link IFluidDataStoreChannel} and {@link IFluidParentContext}, to have a generic specifying `T extends FluidDataStoreMessage` and uses `T["type"]` and `T["content"]` to qualify message related methods, preferably where `submitMessage`, `reSubmit`, and `rollback` have overloads to ensure callers pair values correctly.
- A further improvement would be to reshape `submitMessage`, `reSubmit`, and `rollback` to accept `T` as `message` parameter instead of `type` and `content` parameters that are hard to convince TypeScript must be paired in implementations.

jason-ha committed:
improvement(client): improved internal submitMessage, reSubmit, rollback, and submitSignal type-safety

In Legacy API:
- Add type parameter to `IEnvelope` for `contents` to support stricter typing where desired.
- Deprecate `FluidDataStoreRuntime.submitMessage` as not required per `IFluidDataStoreChannel`. `IFluidParentContext` (which is base interface for `IFluidDataStoreContext`) should always be used to access `submitMessage` functionality.
- Remove unused parameters from mock `submitMessage` and `submitSignal` methods.

Internally:
- Replace `any` in `ISignalEnvelope` with `unknown`.
- Relocate `FluidDataStoreMessage` to runtime-definitions. Eventually this can serve as basis for message types for DataStores.
- In `datastore`, create `LocalFluidDataStoreRuntimeMessage` representing the two known messages for DataStores.
- `FluidDataStoreContext` requires `OutboundFluidDataStoreMessage` for stricter `reSubmit` and `rollback` methods.
- Create `IFluidRootParentContext` for the transition from DataStore to `ContainerRuntime` and use explicit types in `ContainerRuntime` and `ChannelCollection`.
- In `ChannelCollection` restore some old method names with strict typing that internal callers must conform to. The `IFluidDataStoreChannel` portions that were not required by current use were refactored to new `ComposableChannelCollection` class.
- Replace `wrapContext` with `formParentContext` that handles both `IFluidParentContext` and `IFluidRootParentContext` by requiring the specific `submitMessage` and `submitSignal` functions
- Within `ContainerRuntime`:
  - Update `submitMessage` to expect specific messages
  - Update `submitSignal` to expect new `AddressedSignalEnvelope` that is specific subset of `ISignalEnvelope` and is exactly `IEnvelope<ISignalEnvelope["contents"]>`
- Fixup tests to use more realistic data (as required by stricter typing).

Potential Future Improvements:
- Change {@link IFluidDataStoreChannel} and {@link IFluidParentContext}, to have a generic specifying `T extends FluidDataStoreMessage` and uses `T["type"]` and `T["content"]` to qualify message related methods, preferably where `submitMessage`, `reSubmit`, and `rollback` have overloads to ensure callers pair values correctly.
- A further improvement would be to reshape `submitMessage`, `reSubmit`, and `rollback` to accept `T` as `message` parameter instead of `type` and `content` parameters that are hard to convince TypeScript must be paired in implementations.

jason-ha commented on packages/runtime/datastore/src/dataStoreRuntime.ts:
```diff
@@ -1049,8 +1059,12 @@ export class FluidDataStoreRuntime
 		}
 	}
 
+	/**
+	 * @deprecated No implementation required and will be removed in 2.50.
+	 */
 	public submitMessage(type: DataStoreMessageType, content: any, localOpMetadata: unknown) {
```
This deprecation is not required. Just seemed like it might be stale and a clean-up opportunity.

markfields replied:
Yeah I've noticed this isn't used too I think. The asymmetry with submitSignal caught my eye, but I guess the point is, op semantics are internal to the FluidDataStore/DDS implementations, and DataStore authors don't get to generate them manually.  Signals is an open protocol that the runtime makes available to DataStore authors.

markfields replied:
(hence `submitMessage` not existing on `IFluidDataStoreRuntime`)

alexvy86 commented on .changeset/frowny-paths-hear.md:
```diff
@@ -0,0 +1,10 @@
+---
+"@fluidframework/datastore": minor
+---
+---
+"section": legacy
+---
```
We're in a transition period for changesets, this is the new format, with Fluid-specific settings in the same front matter "container", prefixed with double underscore.
```suggestion
---
"@fluidframework/datastore": minor
"__section": legacy
---
```

alexvy86 commented on .changeset/frowny-paths-hear.md:
```diff
@@ -0,0 +1,10 @@
+"section": legacy
+---
+
+Deprecate `FluidDataStoreRuntime.submitMessage`
```
We avoid formatting in headings.
```suggestion
Deprecate FluidDataStoreRuntime.submitMessage
```

alexvy86 commented on .changeset/frowny-paths-hear.md:
```diff
@@ -0,0 +1,10 @@
+
+Deprecate `FluidDataStoreRuntime.submitMessage`
+
+`FluidDataStoreRuntime.submitMessage` implementation is not required per `IFluidDataStoreChannel` and is now deprecated. `IFluidParentContext` (which is base interface for `IFluidDataStoreContext`) should always be used to access `submitMessage` functionality.
```
```suggestion
Implementing `FluidDataStoreRuntime.submitMessage` is not required per `IFluidDataStoreChannel` and is now deprecated. `IFluidParentContext` (the base interface for `IFluidDataStoreContext`) should always be used to access `submitMessage` functionality.
```

alexvy86 reviewed (commented):
Haven't looked at the actual change in detail, but a few doc comments in the meantime

jason-ha committed:
fix(client-test-runtime-utils): restore MockFluidDataStoreRuntime projected types
for `submitSignal` and `submitMessage`.

Additionally, deprecate `MockFluidDataStoreRuntime.submitMessage`.

jason-ha committed:
merge: 'main' into test/client/ops/internal-runtime-type-tightening

markfields commented on packages/runtime/runtime-definitions/api-report/runtime-definitions.legacy.alpha.api.md:
```diff
@@ -133,11 +133,11 @@ export interface IFluidDataStoreChannel extends IDisposable {
     // (undocumented)
     request(request: IRequest): Promise<IResponse>;
-    reSubmit(type: string, content: any, localOpMetadata: unknown): any;
+    reSubmit(type: string, content: any, localOpMetadata: unknown): void;
```
This is technically breaking, right? How are you handling that?

jason-ha replied:
This is not really breaking. API extractor says it was `any` because there was no specification. tsc will see it as untyped which is almost `any`, but no one can use the result. In runtime result is and has always been always `void`. I think we can consider this a documentation level change for accuracy.

markfields commented on packages/runtime/runtime-definitions/src/protocol.ts:
```diff
@@ -126,3 +126,24 @@ export interface IRuntimeMessageCollection {
+ * @internal
+ *
+ * @privateRemarks
+ * Future use opportunity:
```
Sounds nice!

markfields replied:
FYI @anthony-murphy

markfields replied:
Ok I'm talking out of both sides of my mouth - in the other comment I talked down adding these generics, didn't immediately connect the dots to this comment til now. I totally get the structure here and it's obviously appealing in one sense, but I'm concerned about how it will feel in practice.  Maybe too many generics?  Other downsides I'm not thinking of?

markfields commented on packages/runtime/datastore/src/dataStoreRuntime.ts:
```diff
@@ -1136,6 +1149,11 @@ export class FluidDataStoreRuntime
+	 *
+	 * @privateRemarks
+	 * `type` parameter's type of `DataStoreMessageType` is a covariance exception
+	 * over `string` that `IFluidDataStoreChannel` specifies. (`unreachableCase`
```
Some thoughts on this - It's strictly true, but only messages that were submitted from here will come back for resubmit. So IMO it is a legitimate, stable and robust way to type this. I don't think the warning about unreachableCase being hit over time is warranted.

I believe it's correct that `IFluidDataStoreChannel` should specify `string` here, since another implementation could have its own schema. And we haven't wanted to parametrize everything, it leads to quite a soup of generics and doesn't provide much value in practice IMO.

markfields commented on packages/runtime/datastore/src/dataStoreRuntime.ts:
```diff
@@ -1162,8 +1180,12 @@ export class FluidDataStoreRuntime
+	 *
+	 * @privateRemarks
+	 * `type` parameter's type of `DataStoreMessageType` is a covariance exception
+	 * over `string` that `IFluidDataStoreChannel` specifies.
```
Same as above.  If you're going to call this out, please also point out that the only values passed in here at runtime are those passed out via `this.dataStoreContext.submitMessage`

jason-ha replied:
I do believe that is the intent, yet the contract (typing) doesn't say that is definitely the case. It is up to the caller to respect that and not have bugs. I am pretty sure there are improvements that can be made to provide safety and help with readability.
But, yes, we can note the intent in comments in the meantime.

markfields commented on packages/runtime/container-runtime/src/messageTypes.ts:
```diff
@@ -59,63 +60,88 @@ export enum ContainerMessageType {
-	 * Type of the op, within the ContainerRuntime's domain
-	 */
-	type: TType;
+// eslint-disable-next-line @typescript-eslint/no-namespace
```
Talk to me about `namespace`. What does it do, why is the linter saying don't do it, and why is it so rare in our codebase?

jason-ha replied:
I can't get answers to why the linter says don't do it. It may stem from when there was a bunch of churn in the space with ESM modules and confusing terminology.
This was more useful earlier when I was exploring things and it looked like this might leak out. Can and probably should just roll this back.

markfields commented on packages/runtime/container-runtime/src/channelCollection.ts:
```diff
@@ -115,15 +126,47 @@ export const AllowTombstoneRequestHeaderKey = "allowTombstone"; // Belongs in th
+	readonly submitSignal: (envelope: AddressedSignalEnvelope, targetClientId?: string) => void;
 }
 
+type SubmitKeys = "submitMessage" | "submitSignal";
```
Nit: Either export as internal and use in `IFluidRootParentContext` or inline these in both places.

jason-ha replied:
Hmm. There is a separation of concerns here that happens to coincide because of this utility and makes it look like they must be the same. In practice there can be other differences between the two interfaces.
I'll want to look at what is going on with `IFluidParentContextPrivate`. Might be some changes per that.

markfields commented on packages/runtime/container-runtime/src/channelCollection.ts:
```diff
@@ -115,15 +126,47 @@ export const AllowTombstoneRequestHeaderKey = "allowTombstone"; // Belongs in th
+export function formParentContext<T extends IFluidParentContext | IFluidRootParentContext>(
+	context: Omit<IFluidParentContext & IFluidRootParentContext, SubmitKeys>,
+	overrides: Pick<T, SubmitKeys>,
+): Omit<IFluidParentContext & IFluidRootParentContext, SubmitKeys> & Pick<T, SubmitKeys> {
```
Here's a good time to plug this PR of mine I never pushed through #23694  Review comments welcome even as it's closed, I do expect to reopen it.

jason-ha replied:
Mutable should probably go next to Deep/ShallowReadonly in core-interfaces

markfields commented on packages/runtime/container-runtime/src/channelCollection.ts:
```diff
@@ -162,12 +205,8 @@ export function wrapContext(context: IFluidParentContext): IFluidParentContext {
-			return context.submitSignal(...args);
-		},
+		submitMessage: overrides.submitMessage.bind(overrides),
+		submitSignal: overrides.submitSignal,
```
why no `.bind`?

jason-ha replied:
Because there is no need. `submitSignal` is a function and not a method. (We should probably have more functions there we aren't inheriting a class series and overriding.

markfields commented on packages/runtime/container-runtime/src/channelCollection.ts:
```diff
@@ -115,15 +126,47 @@ export const AllowTombstoneRequestHeaderKey = "allowTombstone"; // Belongs in th
+ * @internal
+ * @privateRemarks Exposed per ChannelCollection testing and API extractor request
+ */
+export interface IFluidRootParentContext
```
You'll need to rationalize splitting this out with the new `IFluidParentContextPrivate` added by #24465

markfields commented on packages/runtime/container-runtime/src/channelCollection.ts:
```diff
@@ -242,7 +275,9 @@ export function getLocalDataStoreType(localDataStore: LocalFluidDataStoreContext
  */
-export class ChannelCollection implements IFluidDataStoreChannel, IDisposable {
+export class ChannelCollection
+	implements Omit<IFluidDataStoreChannel, "entryPoint" | "reSubmit" | "rollback">, IDisposable
```
I'm a little wary of this. How will it affect nested datastores? I thought implementing `IFluidDataStoreChannel` was _the_ key feature of this thing, for the purposes of nested datastores.

markfields replied:
Ah, I see `ComposableChannelCollection` below.  Never mind

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -708,11 +711,11 @@ export class ContainerRuntime
 	extends TypedEventEmitter<IContainerRuntimeEvents>
 	implements
 		IContainerRuntime,
+		Omit<IFluidRootParentContext, "submitSignal">,
```
Toss a comment here explaining why the mismatch? (it's a type mismatch that leads you to omit this right?)

jason-ha replied:
Can do. It is not a type mismatch. `ContainerRuntime` doesn't *need* to implement any of `IFluidRootParentContext`. It does implement most of it so that it can hand itself over to the `formParentContext` function to make an `IFluidRootParentContext`. So, the utility here is to force type checking on those members in the class versus there being a type failure when `this` is given to `formParentContext`.

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -1692,26 +1697,29 @@ export class ContainerRuntime
 
-		const parentContext = wrapContext(this);
+		const parentContext = formParentContext<IFluidRootParentContext>(this, {
+			submitMessage: this.submitMessage.bind(this),
```
Nit: A nice improvement would be to support default values for these overrides so you can just omit this here.

jason-ha replied:
I'm not currently in favor of that. Explicit is nice to be clear and have full intent. Alternative would be to have `formRootParentContext` that is just `= formParentContext<IFluidRootParentContext>`. However, there are exactly two of these and one of the `formParentContext<IFluidParentContext>`. Doesn't seem useful. ?

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -4461,7 +4464,7 @@ export class ContainerRuntime
 			case ContainerMessageType.Alias: {
 				// For Operations, call resubmitDataStoreOp which will find the right store
```
Out of date / out of context comment?

jason-ha replied:
Perhaps a little out of date. I suspect this case only handled FluidDataStoreOp to begin with and was expanded for Attach and Alias later.

markfields commented on .changeset/frowny-paths-hear.md:
```diff
@@ -0,0 +1,11 @@
+
+Implementing `FluidDataStoreRuntime.submitMessage` is not required per `IFluidDataStoreChannel` and is now deprecated on `FluidDataStoreRuntime` and corresponding `MockFluidDataStoreRuntime`.
+
+See [issue #24406](https://github.com/microsoft/FluidFramework/issues/24406) for details and alternatives.
```
That issue mentions:

> IFluidParentContext (the base interface for IFluidDataStoreContext) should always be used to access submitMessage functionality.

That may be true for _inside_ `FluidDataStoreRuntime` class, but as for the public API to submit an op, I think the comment should just be - "Submitting arbitrary messages is not supported"

jason-ha replied:
Is `FluidDataStoreRuntime` a class customers are expected to derive from? I don't think Pages does and I see that we do internally in at least two places.
Feel free edit the issue.
We should seal the class if no one should derive from it. Then it will be easier to blanketly say don't call.

markfields reviewed (approved):
LGTM overall! I'll take another pass once you're all merged up and have made the changes you're mentioning in comments

jason-ha committed:
merge: 'main' into test/client/ops/internal-runtime-type-tightening

jason-ha committed:
merge: 'main' into test/client/ops/internal-runtime-type-tightening

jason-ha committed:
improvement(client): cleanup post merge

jason-ha committed:
doc(client): additional/fixed comments per PR feedback

jason-ha committed:
build(client-container-runtime): revert namespace use

jason-ha committed:
merge: 'main' into test/client/ops/internal-runtime-type-tightening

jason-ha committed:
merge: 'main' into test/client/ops/internal-runtime-type-tightening

Josmithr commented on packages/runtime/container-runtime/src/channelCollection.ts:
```diff
@@ -128,20 +145,44 @@ interface FluidDataStoreMessage {
-export function wrapContext(context: IFluidParentContextPrivate): IFluidParentContextPrivate {
+export interface IFluidRootParentContextPrivate
+	extends Omit<IFluidParentContextPrivate, "submitMessage" | "submitSignal"> {
+	readonly submitMessage: (
```
Nit: probably worth documenting the method variants here, since intellisense won't (or at least shouldn't) pull them from `IFluidParentContextPrivate`.

Josmithr reviewed (approved):
API changes look good, assuming we're in the window of time where we can make beta breaks.

jason-ha committed:
merge: 'main' into test/client/ops/internal-runtime-type-tightening

jason-ha commented on packages/runtime/test-runtime-utils/api-report/test-runtime-utils.legacy.beta.api.md:
```diff
@@ -502,7 +502,7 @@ export class MockFluidDataStoreRuntime extends EventEmitter implements IFluidDat
     setConnectionState(connected: boolean, clientId?: string): void;
     // (undocumented)
-    submitSignal(type: string, content: any): null;
+    submitSignal: IFluidDataStoreRuntime["submitSignal"];
```
Is it unfortunate?
The signature for `IFluidDataStoreRuntime.submitSignal` was changes to `(type: string, content: unknown, targetClientId?: string) => void` 19 months ago. There was no visit here. This will keep its type up to date.

jason-ha replied:
Gotcha. Note that there was no reported change before for at least some cases of breaking change to `IFluidDataStoreRuntime.submitSignal`.
I think the IFluidDataStoreRuntime.submitSignal change should be the larger signal to anything in this space.

jason-ha commented on packages/test/local-server-tests/src/test/opsOnReconnect.spec.ts:
```diff
@@ -135,19 +139,20 @@ describe("Ops on Reconnect", () => {
-					const address = envelope.contents.content.address;
-					const content = envelope.contents.content.contents;
+					const envelope = message.contents as IEnvelope<FluidDataStoreMessage>;
+					// explicitly using `any` even if IEnvelope default is later changed to unknown.
```
I don't think there is value doing that. I can revise the comment.

jason-ha committed:
docs(client): additions per feedback

jason-ha committed:
revert(client-datastore): keep `rollback` looking optional

jason-ha committed:
merge: 'main' into test/client/ops/internal-runtime-type-tightening
